### PR TITLE
Add example : `Data Parallel Training with KerasNLP`

### PR DESCRIPTION
**Pull request by shivance**
Closes : [keras-nlp/#858](https://github.com/keras-team/keras-nlp/issues/858)

This tutorial demonstrates how to perform distributed training using KerasNLP.  It covers the steps for training a BERT-based masked language model on the wikitext-2 
dataset, using multiple GPUs for accelerated training.

cc : @chenmoneygithub & @mattdangerw

### CODE REVIEW

File: examples/nlp/distributed_training_with_keras_nlp.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    tf.keras.callbacks.TensorBoard(log_dir='./logs'),
+    tf.keras.callbacks.LearningRateScheduler(decay),
+    PrintLR()
+]
+
+print(tf.config.list_physical_devices('GPU'))
+strategy = tf.distribute.MirroredStrategy()
+print('Number of devices: {}'.format(strategy.num_replicas_in_sync))
+
+with strategy.scope():
```
Comment (mattdangerw) [inline examples/nlp/distributed_training_with_keras_nlp.py:114]: this might be a good place to add more text on what a distribution strategy scope does, and why the model needs to be created in scope

File: examples/nlp/distributed_training_with_keras_nlp.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+training.
+
+In this tutorial, we will use KerasNLP to train a BERT-based masked language model (MLM) 
+on the wikitext-2 dataset. The MLM task involves predicting the masked words in a sentence, 
+which helps the model learn contextual representations of words.
+
+## Setup
+
+The tutorial relies on KerasNLP 0.5.2. Additionally, we need
+at least TensorFlow 2.11 in order to use AdamW with mixed precision.
```
Comment (mattdangerw) [inline examples/nlp/distributed_training_with_keras_nlp.py:30]: maybe worth introducing mixed precision in just a sentence. e.g. explain why you should basically always turn it on

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Thanks! Great to have a simple focused guide like this. Dropped some initial comments.

**Comment by mattdangerw** [comment]
Actually re the title. Because we are going to have a few distributed guides. Let's change the title to...

`Data Parallel Training with KerasNLP`

File: examples/nlp/distributed_training_with_keras_nlp.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        print('\nLearning rate for epoch {} is {}'.format(epoch + 1, model_dist.optimizer.lr.numpy()))
+
+# Put all the callbacks together.
+callbacks = [
+    tf.keras.callbacks.TensorBoard(log_dir='./logs'),
+    tf.keras.callbacks.LearningRateScheduler(decay),
+    PrintLR()
+]
+
+print(tf.config.list_physical_devices('GPU'))
```
[the remark this answers is not in the document]
Comment (shivance) [reply examples/nlp/distributed_training_with_keras_nlp.py:110]: Hey @mattdangerw I found that this tutorial demonstrates DTensor API and training involved with it, at the core.
Isn't it different from our goal with this tutorial?

File: examples/nlp/distributed_training_with_keras_nlp.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+"""
+Title: Distributed Training with KerasNLP
+Author: Anshuman Mishra
+Date created: 2023/06/04
+Last modified: 2023/06/05
+Description: This tutorial demonstrates how to perform distributed training using KerasNLP. 
+It covers the steps for training a BERT-based masked language model on the wikitext-2 
+dataset, using multiple GPUs for accelerated training.
+
+Accelerator: GPU
```
Comment (shivance) [inline examples/nlp/distributed_training_with_keras_nlp.py:11]: Also we assume our default accelerator to be GPU here rather than setting up Logical Devices from CPU

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    origin="https://s3.amazonaws.com/research.metamind.io/wikitext/wikitext-2-v1.zip",
+    extract=True,
+)
+wiki_dir = os.path.expanduser("~/.keras/datasets/wikitext-2/")
+
+# Load wikitext-103 and filter out short lines.
+wiki_train_ds = (
+    tf.data.TextLineDataset(wiki_dir + "wiki.train.tokens")
+    .filter(lambda x: tf.strings.length(x) > 100)
+    .batch(PRETRAINING_BATCH_SIZE)
```
Comment (chenmoneygithub) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:91]: Let's add the best practice here:
```
.batch(PRETRAINING_BATCH_SIZE).cache().prefetch(tf.data.AUTOTUNE)
```

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+In the above code, we download the wikitext-2 dataset and extract it. Then, we define 
+three datasets: wiki_train_ds, wiki_val_ds, and wiki_test_ds. These datasets are 
+filtered to remove short lines and are batched for efficient training.
+
+Let's define a function for decaying the learning rate. You can define any decay function you need.
+We also define a callback for printing the learning rate at the end of each epoch.
+"""
+
+
+def decay(epoch):
```
Comment (chenmoneygithub) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:114]: Any specific reason for using a per-epoch decay compared to a normal `LearningRateSchedule`? The latter approach is more commonly used IMO

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        return 1e-3
+    elif epoch >= 3 and epoch < 7:
+        return 1e-4
+    else:
+        return 1e-5
+
+
+class PrintLR(tf.keras.callbacks.Callback):
+    def on_epoch_end(self, epoch, logs=None):
+        print(
```
Comment (chenmoneygithub) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:125]: Let's print with f-string for cleaner code

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+With the datasets prepared, we now initialize and compile our model and optimizer within
+the `strategy.scope()`:
+"""
+with strategy.scope():
+    # Everything that creates variables should be under the strategy scope.
+    # In general this is only model construction & `compile()`.
+    model_dist = keras_nlp.models.BertMaskedLM.from_preset("bert_tiny_en_uncased")
+    model_dist.compile(
+        loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+        optimizer=keras.optimizers.Adam(5e-5),
```
Comment (chenmoneygithub) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:172]: This learning rate is contradicting with the callback defined above. Basically you start with 5e-5, then at the end of first epoch it's bumped up to 1e-3.

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+        weighted_metrics=keras.metrics.SparseCategoricalAccuracy(),
+    )
+
+"""
+Let's train our model
```
Comment (chenmoneygithub) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:177]: Nit: period at the end.

**Review by chenmoneygithub — CHANGES_REQUESTED** [review]
Thanks for the PR! Generally well written, left a few comments

File: examples/nlp/data_parallel_training_with_keras_nlp.py (hunk 6 above)
Comment (shivance) [reply examples/nlp/data_parallel_training_with_keras_nlp.py:114]: @chenmoneygithub  On a second thought, I think we should remove the learning rate decay. Focusing only on the data parallel training demonstration.
Other keras tutorials better demonstrate using LR decay to improve accuracy.

Comment (chenmoneygithub) [reply examples/nlp/data_parallel_training_with_keras_nlp.py]: Let's throw a simple learning rate schedule:
```
lr = tf.keras.optimizers.schedules.PolynomialDecay(
    5e-5,
    decay_steps={total_training_steps},
    end_learning_rate=0.0,
)
```

It's a common practice to use a decayed learning rate in NLP training/tuning. A callback is a bit over complex, but this should work.

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+## Introduction
+
+Distributed training is a technique used to train deep learning models on multiple devices 
+or machines simultaneously. It helps to reduce training time and improve model performance 
+by leveraging the computational power of multiple resources. KerasNLP is a library that 
+provides tools and utilities for natural language processing tasks, including distributed 
+training.
+
+In this tutorial, we will use KerasNLP to train a BERT-based masked language model (MLM) 
+on the wikitext-2 dataset. The MLM task involves predicting the masked words in a sentence,
```
Comment (mattdangerw) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:23]: maybe add `(a 2 million word dataset of wikipedia articles)` so people know what this is.

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+
+Specifically, this guide teaches you how to use the `tf.distribute` API to train Keras
+models on multiple GPUs, with minimal changes to your code, in the following two setups:
+
+- On multiple GPUs (typically 2 to 8) installed on a single machine (single host,
+multi-device training). This is the most common setup for researchers and small-scale
+industry workflows.
+- On a cluster of many machines, each hosting one or multiple GPUs (multi-worker
+distributed training). This is a good setup for large-scale industry workflows, e.g.
+training high-resolution image classification models on tens of millions of images using
```
Comment (mattdangerw) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:39]: can re rephrase this in terms on text not images? as this is NLP guide?

maybe "billion word datasets on 20-100 GPUs."

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+or machines simultaneously. It helps to reduce training time and improve model performance 
+by leveraging the computational power of multiple resources. KerasNLP is a library that 
+provides tools and utilities for natural language processing tasks, including distributed 
+training.
+
+In this tutorial, we will use KerasNLP to train a BERT-based masked language model (MLM) 
+on the wikitext-2 dataset. The MLM task involves predicting the masked words in a sentence, 
+which helps the model learn contextual representations of words. 
+
+This guide focuses on data parallelism, in particular synchronous data parallelism, where
```
Comment (mattdangerw) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:26]: maybe reword this a little

This guide focuses on data parallelism, in particular synchronous data parallelism, where each accelerator (a GPU or TPU) holds a complete replica of the model, and sees a different partial batch of the input data. Partial gradients are computed on each device, aggregated, and used to compute a global gradient update.

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+20-100 GPUs.
+
+## Setup
+
+The tutorial relies on KerasNLP 0.5.2. Additionally, we need
+at least TensorFlow 2.11 in order to use AdamW with mixed precision.
+"""
+
+"""shell
+pip install keras-nlp==0.5.2 -q
```
Comment (mattdangerw) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:49]: I would not pin here. Let's just install `keras-nlp` (whatever is the latest). We try to test these guides when updating our library.

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+    tf.keras.callbacks.TensorBoard(log_dir="./logs"),
+    PrintLR(),
+]
+
+
+print(tf.config.list_physical_devices("GPU"))
+
+"""
+To do single-host, multi-device synchronous training with a Keras model, you would use
+the [`tf.distribute.MirroredStrategy` API](
```
Comment (mattdangerw) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:138]: I think if you do `tf.distribute.MirroredStrategy` (without a link), keras.io will generate it automatically for you

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+Let's also make a callback to TensorBoard, this will enable visualization of different metrics
+while we train the model in later part of this tutorial We put all the callbacks together as follows:
+"""
+callbacks = [
+    tf.keras.callbacks.TensorBoard(log_dir="./logs"),
+    PrintLR(),
+]
+
+
+print(tf.config.list_physical_devices("GPU"))
```
Comment (mattdangerw) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:134]: This will only show a single GPU right on colab right? we might actually want to use this https://www.tensorflow.org/api_docs/python/tf/config/set_logical_device_configuration

You can get the total memory to partition with
https://www.tensorflow.org/api_docs/python/tf/config/experimental/get_memory_info

That will help show a proper two device setup, on a single colab, which I think will be useful to people

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+With the datasets prepared, we now initialize and compile our model and optimizer within
+the `strategy.scope()`:
+"""
+with strategy.scope():
+    # Everything that creates variables should be under the strategy scope.
+    # In general this is only model construction & `compile()`.
+    model_dist = keras_nlp.models.BertMaskedLM.from_preset("bert_tiny_en_uncased")
+    model_dist.compile(
+        loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+        optimizer=tf.keras.optimizers.Adam(5e-5),
```
Comment (mattdangerw) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:163]: You mention AdamW above, but then use just Adam here, should we show AdamW?

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+        weighted_metrics=keras.metrics.SparseCategoricalAccuracy(),
+    )
+
+"""
+Let's train our model.
```
Comment (mattdangerw) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:168]: maybe a little more color here.

"After creating our model under the scope, fit will automatically run distributed training. Just call it normally!"

**Review by mattdangerw — APPROVED** [review]
Nice! Just a few comments.

**Comment by fchollet** [comment]
Looking good! Please add the generated files.

**Comment by shivance** [comment]
Hi @mattdangerw , I've added the logical devices snippet, also added a lr schedule as suggested by @chenmoneygithub. Could you review once more?

**Comment by shivance** [comment]
I've added generated files @fchollet.

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 18]
```diff
@@ -0,0 +1,221 @@
+"""
+Title: Data Parallel Training with KerasNLP
+Author: Anshuman Mishra
+Date created: 2023/06/04
+Last modified: 2023/06/05
```
Comment (fchollet) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:5]: Update date (date created should be roughly the publication date)

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+Description: Data Parallel training with KerasNLP.
+Accelerator: GPU
+"""
+
+"""
+## Introduction
```
Comment (fchollet) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:11]: Great intro!

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 20]
```diff
@@ -0,0 +1,221 @@
...
+"""
+
+keras.utils.get_file(
+    origin="https://s3.amazonaws.com/research.metamind.io/wikitext/wikitext-2-v1.zip",
+    extract=True,
+)
+wiki_dir = os.path.expanduser("~/.keras/datasets/wikitext-2/")
+
+# Load wikitext-103 and filter out short lines.
+wiki_train_ds = (
```
Comment (fchollet) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:98]: The data is not shuffled, is that intended?

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 21]
```diff
@@ -0,0 +1,221 @@
...
+EPOCHS = 3
+
+"""
+It's a common practice to use a decayed learning rate in NLP training/tuning. We'll
+use `PolynomialDecay` schedule here.
+"""
+
+total_training_steps = sum(1 for _ in wiki_train_ds.as_numpy_iterator()) * EPOCHS
+lr_schedule = tf.keras.optimizers.schedules.PolynomialDecay(
+    5e-5,
```
Comment (fchollet) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:135]: Please use a named keyword arg for better readability

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 22]
```diff
@@ -0,0 +1,221 @@
...
+print("Number of devices: {}".format(strategy.num_replicas_in_sync))
+
+"""
+With the datasets prepared, we now initialize and compile our model and optimizer within
+the `strategy.scope()`:
+"""
+
+with strategy.scope():
+    # Everything that creates variables should be under the strategy scope.
+    # In general this is only model construction & `compile()`.
```
Comment (fchollet) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:183]: It's a good practice to also include `fit()` because `fit()` upon first call might create variables

**Review by fchollet — COMMENTED** [review]
Very nice work!

File: examples/nlp/data_parallel_training_with_keras_nlp.py (hunk 19 above)
Comment (shivance) [reply examples/nlp/data_parallel_training_with_keras_nlp.py:11]: Thanks !

File: examples/nlp/data_parallel_training_with_keras_nlp.py (hunk 22 above)
Comment (shivance) [reply examples/nlp/data_parallel_training_with_keras_nlp.py:183]: Noted !

**Comment by shivance** [comment]
Will add updated generated files soon !

**Comment by shivance** [comment]
Adding a filter warning !

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 23]
```diff
@@ -0,0 +1,221 @@
...
+Description: Data Parallel training with KerasNLP.
+Accelerator: GPU
+"""
+
+"""
+## Introduction
+
+Distributed training is a technique used to train deep learning models on multiple devices
+or machines simultaneously. It helps to reduce training time and allows for training larger
+models with more data.KerasNLP is a library that provides tools and utilities for natural
```
Comment (abheesht17) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:15]: Space after full-stop (before "KerasNLP")

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 24]
```diff
@@ -0,0 +1,221 @@
...
+import os
+import tensorflow as tf
+from tensorflow import keras
+import keras_nlp
+
+"""
+Before we start any training, let's configure our single GPU to show up as two logical
+devices.
+
+When you are training with two or more phsyical GPUs, this is totally uncessary. This
```
Comment (abheesht17) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:61]: uncessary --> unnecessary

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 25]
```diff
@@ -0,0 +1,221 @@
...
+is just a trick to show real distributed training on the default colab GPU runtime,
+which has only one GPU availabe.
```
Comment (abheesht17) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:63]: availabe --> available

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 26]
```diff
@@ -0,0 +1,221 @@
...
+"""
+
+physical_devices = tf.config.list_physical_devices("GPU")
+tf.config.set_logical_device_configuration(
+    physical_devices[0],
+    [
+        tf.config.LogicalDeviceConfiguration(memory_limit=15360 // 2),
+        tf.config.LogicalDeviceConfiguration(memory_limit=15360 // 2),
+    ],
+)
```
Comment (abheesht17) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:77]: Nice! Didn't know we could do this.

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 27]
```diff
@@ -0,0 +1,221 @@
...
+class PrintLR(tf.keras.callbacks.Callback):
+    def on_epoch_end(self, epoch, logs=None):
+        print(
+            f"\nLearning rate for epoch {epoch + 1} is {model_dist.optimizer.lr.numpy()}"
+        )
+
+
+"""
+Let's also make a callback to TensorBoard, this will enable visualization of different
+metrics while we train the model in later part of this tutorial We put all the callbacks
```
Comment (abheesht17) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:155]: Full-stop before "We"

File: examples/nlp/data_parallel_training_with_keras_nlp.py [hunk 28]
```diff
@@ -0,0 +1,221 @@
...
+- Instantiate a `MirroredStrategy`, optionally configuring which specific devices you
+want to use (by default the strategy will use all GPUs available).
+- Use the strategy object to open a scope, and within this scope, create all the Keras
+objects you need that contain variables. Typically, that means **creating & compiling the
+model** inside the distribution scope.
+- Train the model via `fit()` as usual.
+"""
+
+strategy = tf.distribute.MirroredStrategy()
+print("Number of devices: {}".format(strategy.num_replicas_in_sync))
```
Comment (abheesht17) [inline examples/nlp/data_parallel_training_with_keras_nlp.py:179]: .format() is generally not recommended in Python 3.x. You can use f-string here.

```
print(f"Number of devices: {strategy.num_replicas_in_sync}")
```

**Review by abheesht17 — CHANGES_REQUESTED** [review]
Cool stuff! Left some small comments

File: examples/nlp/data_parallel_training_with_keras_nlp.py (hunk 26 above)
Comment (shivance) [reply examples/nlp/data_parallel_training_with_keras_nlp.py:77]: :D

File: examples/nlp/data_parallel_training_with_keras_nlp.py (hunk 28 above)
Comment (shivance) [reply examples/nlp/data_parallel_training_with_keras_nlp.py:179]: Done !

**Review by fchollet — APPROVED** [review]
LGTM, thanks for the great contribution!
